sparkle/src/components/ElementModal.tsx

```diff
@@ -0,0 +1,26 @@
+import React, { useState } from "react";
+
+import { Modal, ModalProps } from "./Modal";
+
+type ElementModalProps<T> = Omit<ModalProps, "isOpen"> & {
+  openOnElement: T | null;
+};
+
+export function ElementModal<T>({
+  openOnElement,
+  ...props
+}: ElementModalProps<T>) {
+  const { onClose } = props;
+  const [isClosingTransition, setIsClosingTransition] = useState(false);
+  const realOnClose = () => {
+    setIsClosingTransition(true);
+    setTimeout(() => {
+      onClose();
+      setIsClosingTransition(false);
+    }, 200);
+  };
```

Comment:
We should not mix business logic with rendering needs (meaning not delaying `onClose` for the sake of having great transitions). IMHO, we should rely on a `useEffect()` that proxy the `openOnElement` to the `<Modal>` component below.

Example flow:
When `openOnElement` changes, we go through the `useEffect` which keeps a copy of the value passed to the modal and only change it after the transition is done.

Reply:
Unfortunately as evoked IRL it would not work, since the modal transition for closing happens when the isOpen prop changes to false, and this cannot happen before the openOnElement is set to null

I would agree mixing business logic and rendering logic is not optimal. In modals however, the onClose should be only rendering logic: the modal cannot close itself (closing is a form of rendering) so the parent gives the modal a way to do it; 

The onClose always have a rendering purpose first, as opposed to the onSave which is another matter. The misuse would be to put business logic in the onClose IMHO. Since it's likely we'll do it anyway, we could either document, or make it more explicit in code. Make a modal interface that clearly separate the code needed for rendering (onclose) and the code related to business logic

Reply:
Seen IRL, let do:

```
import React, { useMemo, useState } from "react";

import { Modal, ModalProps } from "./Modal";

type ElementModalProps<T> = Omit<ModalProps, "isOpen"> & {
  openOnElement: T | null;
};

export function ElementModal<T>({
  openOnElement,
  onClose,
  ...props
}: ElementModalProps<T>) {
  const [isClosingTransition, setIsClosingTransition] = useState(false);

  const realOnClose = useMemo(
    () => () => {
      setIsClosingTransition(true);
      setTimeout(() => {
        onClose();
        setIsClosingTransition(false);
      }, 200);
    },
    [onClose]
  );

  return (
    <Modal
      isOpen={openOnElement !== null && !isClosingTransition}
      onClose={realOnClose}
      {...props}
    />
  );
}
```